File: backend/app/services/access_roster.py
"""THE SIGN-IN GATE: who this institution lets in, and what a person who is not let in is told.

This application shipped for months with no sign-in refusal of any kind. Anyone holding a Google
account whose address Google would vouch for got a `User` row and a bearer token, automatically, at
``DEFAULT_SIGNUP_ROLE``. This module is the end of that, and every consequence in here is new.

FOUR THINGS THIS MODULE OWNS, AND THE ARGUMENT FOR EACH.

1. **``roster_admits`` — the one question the gate asks.** It tests ``status = ACTIVE`` in the WHERE
   clause, not the existence of a row, and that distinction is the whole of the security model. A
   PENDING row is created BY THE REFUSED CALLER; if "a row exists" admitted anybody, then the first
   refusal would mint the credential for the second attempt. Anyone refactoring this into
   ``find_unique(where={"email": ...}) is not None`` has written an authentication bypass that the
   attacker triggers on their own address, in one HTTP request, with no privileges at all.

2. **``record_access_request`` — the bounded, unauthenticated write.** See its docstring. In short:
   the caller must have PROVED they own the address before a row is written, one address is one row
   forever, the only attacker-controlled value stored is the address itself, and the table has a
   ceiling.

3. **The refusal sentences.** The product owner ruled explicitly that "awaiting approval" and "wrong
   email or password" must be DIFFERENT answers. That widens account enumeration on purpose — a
   stranger can learn that an address is known to this institution — and the trade was made
   deliberately, because the alternative is a person who is waiting on an administrator spending a
   week resetting a password that was never wrong. What the refusals must NOT do is leak anything
   past that single fact: no name, no role, no whether a password was ever set, no hint about any
   other account. Read ``refusal_detail`` before changing a word of them.

4. **MASTER_ADMIN IS NEVER GATED.** ``is_master_admin_email`` is consulted before the roster on
   every path. The roster is a table only an administrator can edit; an administrator locked out by
   it is an outage with no remedy inside the product, because there is nobody left who can add the
   row that lets anybody — including them — back in. The exemption is keyed on the CONFIGURED
   address rather than on ``User.role`` so that it works even when the account does not exist yet,
   which is precisely the situation a break-glass has to survive.

EVERY EMAIL THAT REACHES THIS MODULE IS LOWER-CASED BY :func:`normalise_email`, and by nothing else.
``AccessRoster.email`` is unique and is joined to ``User.email``; a roster row typed
``A.Sharma@Example.org`` that never matches the account signing in as ``a.sharma@example.org`` is a
person locked out of the application by a capital letter, with an administrator looking straight at
the row that was supposed to let them in.
"""

import logging
from datetime import UTC, datetime
from typing import Any

from app.core.config import get_settings
from app.core.db import db

logger = logging.getLogger(__name__)

#: The only status that admits. Named so that reading the gate does not require remembering which of
#: the four enum values is the permissive one.
ADMITTING_STATUS = "ACTIVE"

#: Statuses an administrator may set from the roster screen. PENDING is absent on purpose: "put this
#: request back in the queue" is not an action anybody asked for, and it would let one admin quietly
#: undo another's rejection in a way that looks, on the queue, exactly like a fresh request from the
#: applicant. Restoring a rejected person is an approval — ACTIVE — which is an act with a name and
#: a ``decidedById`` against it.
DECIDABLE_STATUSES = frozenset({"ACTIVE", "REJECTED", "SUSPENDED"})

#: An email longer than this is not an email. RFC 5321 caps a forward path at 254 characters, and
#: the column is unbounded TEXT, so this is the thing standing between the pending queue and a row
#: whose primary content is a megabyte of an attacker's choosing. Over-length addresses are dropped
#: rather than truncated: a truncated address is a DIFFERENT address, and storing it would put a
#: row in the admin's queue for a person who does not exist.
MAX_EMAIL_LENGTH = 254

#: Caps on the two free-text columns an ADMINISTRATOR writes. Not a security boundary — an admin can
#: already write anything they like all over this application — but the roster is rendered in a
#: table on two clients, and a note the length of a dissertation makes the queue unreadable for
#: every other admin.
MAX_FULL_NAME_LENGTH = 200
MAX_NOTES_LENGTH = 2000


# ---------------------------------------------------------------------------------------------
# The refusals
# ---------------------------------------------------------------------------------------------
#
# THESE TWO SENTENCES ARE THE FEATURE. Everything else in this module exists to decide which one a
# person reads.
#
# The wrong-credential answer is UNCHANGED from what this application has always said, and it is
# still a 401. Do not "improve" it into "no account with that address exists" — the enumeration the
# product owner accepted is specifically about the ROSTER's opinion of an address, not about a free
# oracle for whether a password is right.
INVALID_CREDENTIALS_DETAIL = "Invalid email or password"

# THE PENDING SENTENCE. Three jobs, in this order: say the request exists (so the person stops
# retyping their password), say who is going to act on it (so they stop emailing the wrong desk),
# and say that nothing further is required of them (so they stop re-submitting, which is the
# behaviour that fills the admin's queue with noise).
#
# "SIGN-IN DETAILS", NOT "PASSWORD". The identical sentence is served to the Google branch, where
# there is no password at all — a refusal that told a Google user their password was fine would be
# saying something false about a thing they never typed, and the first support call it produces is
# somebody asking which password the app means.
ACCESS_PENDING_DETAIL = (
    "Your access request is awaiting approval by an administrator. "
    "Your sign-in details are correct; there is nothing more to do until an administrator approves "
    "this address."
)

# THE REJECTED SENTENCE. It does not say "an administrator rejected you", and that is a considered
# choice rather than squeamishness: the person cannot appeal to this screen, and the only thing that
# changes their situation is talking to a human. Saying "not approved" states the fact, and naming
# the administrator names the one action that leads anywhere. It is deliberately DISTINCT from the
# pending sentence, because telling a rejected person they are "awaiting approval" is a lie that
# guarantees they wait forever.
ACCESS_REJECTED_DETAIL = (
    "This address is not approved for access to the repository. "
    "Contact your administrator if you believe this is a mistake."
)

# THE SUSPENDED SENTENCE. Distinct again, because this person HAD access and it was ended — telling
# them they are "awaiting approval" would have them waiting for something that already happened, and
# telling them "not approved" hides that they are looking at a revocation they may be able to get
# reversed today.
ACCESS_SUSPENDED_DETAIL = (
    "Your access to the repository has been suspended. Contact your administrator."
)

#: Status -> the sentence that status is refused with. A status missing from here would fall back to
#: the rejected wording, which is the safe direction: it never invites somebody to wait.
_REFUSAL_DETAILS: dict[str, str] = {
    "PENDING": ACCESS_PENDING_DETAIL,
    "REJECTED": ACCESS_REJECTED_DETAIL,
    "SUSPENDED": ACCESS_SUSPENDED_DETAIL,
}


def refusal_detail(status: Any) -> str:
    """The sentence a person in *status* reads. Never returns the wrong-credential wording."""
    return _REFUSAL_DETAILS.get(status_value(status), ACCESS_REJECTED_DETAIL)


# ---------------------------------------------------------------------------------------------
# Normalisation
# ---------------------------------------------------------------------------------------------


def normalise_email(email: Any) -> str:
    """The one canonical form of an email address in this system. See the module docstring.

    Returns ``""`` for anything absent, blank or over-length; every caller treats the empty string
    as "no usable address", which refuses rather than admits.
    """
    address = str(email or "").strip().lower()
    if not address or len(address) > MAX_EMAIL_LENGTH:
        return ""
    return address


def status_value(status: Any) -> str:
    """A prisma enum member, or a plain string, as a plain string."""
    return str(getattr(status, "value", status) or "")


def role_value(role: Any) -> str:
    return str(getattr(role, "value", role) or "")


def is_master_admin_email(email: Any) -> bool:
    """Is this THE break-glass address? See point 4 of the module docstring.

    Keyed on the configured ``MASTER_ADMIN_EMAIL`` and not on ``User.role``, so it holds before the
    account exists, after somebody has deleted the row, and while the roster says otherwise.
    """
    address = normalise_email(email)
    if not address:
        return False
    return address == normalise_email(get_settings().master_admin_email)


def gate_enforced() -> bool:
    """Is the gate switched on? See ``Settings.access_roster_enforced`` for why the switch exists."""
    return bool(get_settings().access_roster_enforced)


# ---------------------------------------------------------------------------------------------
# The gate
# ---------------------------------------------------------------------------------------------


async def roster_row(email: Any) -> Any | None:
    """The roster row for an address, or ``None``. Does not decide anything; ``roster_admits`` does."""
    address = normalise_email(email)
    if not address:
        return None
    return await db.accessroster.find_unique(where={"email": address})


async def roster_admits(email: Any) -> bool:
    """Is there an ACTIVE roster row for this address?

    THE ``status`` TEST IS IN THE QUERY, NOT IN PYTHON, and it is not an optimisation. See point 1
    of the module docstring: a row exists for people who have been refused, so "found a row" and
    "the institution admits them" are one clause apart and conflating them is an authentication
    bypass an unauthenticated caller can arrange for themselves.
    """
    address = normalise_email(email)
    if not address:
        return False
    row = await db.accessroster.find_first(
        where={"email": address, "status": ADMITTING_STATUS}
    )
    return row is not None


async def mark_first_seen(email: Any) -> None:
    """Stamp ``firstSeenAt`` the first time an admitted address actually gets in.

    Written ONCE — the WHERE clause carries ``firstSeenAt: None`` — so the stamp records the first
    sign-in rather than the most recent one, and two simultaneous logins cannot race each other into
    overwriting it. Silent when there is no row (the master admin's exemption means they can sign in
    without one); an ``update_many`` that matches nothing is cheaper than deciding beforehand whether
    to ask, on a path a person takes once a week.
    """
    address = normalise_email(email)
    if not address:
        return
    await db.accessroster.update_many(
        where={"email": address, "firstSeenAt": None},
        data={"firstSeenAt": datetime.now(UTC)},
    )


async def record_access_request(email: Any) -> None:
    """Record that somebody who PROVED they own *email* was refused. The unauthenticated write.

    THE CALLER MUST HAVE PROVED OWNERSHIP OF THE ADDRESS BEFORE CALLING THIS. In ``auth.py`` that
    means one of exactly two things: a Google ID token this server verified, carrying
    ``email_verified``; or an existing local account whose password hash the supplied password
    matched. A bare "unknown address + any password" attempt writes NOTHING, because it proves
    nothing — that path is the one an anonymous script can drive at line rate, and letting it seed
    the administrators' queue would turn the notification this feature exists to provide into spam
    that hides the real requests. THAT ORDERING IS THE PRIMARY BOUND ON THIS WRITE. Moving this call
    earlier in the sign-in flow, to "capture more requests", removes it.

    THE OTHER BOUNDS, EACH OF WHICH IS LOAD-BEARING:

    * **One address is one row, forever.** A repeat attempt is an UPDATE of ``requestCount`` and
      ``lastRequestedAt``, never an INSERT. Somebody hammering the form produces one row with a
      rising counter, which is also exactly what an administrator needs to see.
    * **Nothing attacker-controlled is stored beyond the address.** No display name, no picture, no
      "reason for request" field — not because those would be useless to an admin, but because the
      pending queue is the only screen in this product where a stranger can cause content to appear,
      and the first thing that goes wrong with such a screen is that it renders somebody's chosen
      text to an administrator. ``fullName`` and ``notes`` are ADMIN-ONLY columns. Do not add a
      field here "just for context".
    * **The address itself is length-capped** by ``normalise_email`` before it reaches the database.
    * **The table has a ceiling** (``ACCESS_ROSTER_MAX_PENDING``). At the cap, a request for an
      address not already on the roster is DROPPED — the person is still refused, with the identical
      sentence, so nothing about the cap is observable to them; only the log and the admin's count
      say so. Dropping is the right failure: the alternative is an unbounded table written by
      anonymous callers, and an admin cannot work a queue of fifty thousand entries anyway.

    A REJECTED PERSON DOES NOT SILENTLY RE-QUEUE. Their re-attempt bumps the counter and the
    timestamp and LEAVES THE STATUS AT REJECTED. Three options were on the table — re-open, stay
    rejected, or rate-limit the re-ask — and re-opening is the one that makes the queue unworkable:
    a rejection an applicant can undo by clicking "sign in with Google" again is not a rejection, it
    is a timer, and the administrator gets the same request back every day forever with no way to
    make it stop. Staying rejected keeps the decision a decision; the rising ``requestCount`` is how
    an admin notices somebody who keeps trying and goes to talk to them. The same argument applies
    to SUSPENDED, which must obviously not be undone by the suspended person retrying. ACTIVE rows
    never reach here at all — they were admitted.
    """
    address = normalise_email(email)
    if not address:
        return

    now = datetime.now(UTC)
    existing = await db.accessroster.find_unique(where={"email": address})
    if existing is not None:
        # The status is deliberately NOT touched. See the last paragraph of the docstring.
        await db.accessroster.update(
            where={"email": address},
            data={
                "requestCount": (existing.requestCount or 0) + 1,
                "lastRequestedAt": now,
            },
        )
        return

    cap = max(0, int(get_settings().access_roster_max_pending))
    pending = await pending_count()
    if pending >= cap:
        logger.warning(
            "Access request from a new address dropped: the pending roster is at its ceiling "
            "(%s of %s). An administrator must clear the queue before new requests are recorded.",
            pending,
            cap,
        )
        return

    await db.accessroster.create(
        data={
            "email": address,
            "status": "PENDING",
            # grantedRole is left at the column default — the lowest rung. "All the users by default
            # join as the lowest rung unless promoted there itself": the promotion happens on the
            # roster screen, at approval time, and never here.
            "requestCount": 1,
            "firstRequestedAt": now,
            "lastRequestedAt": now,
        }
    )


async def pending_count() -> int:
    """How many addresses are waiting on an administrator. THE NOTIFICATION.

    There is no email and no push in this codebase — neither has ever existed in it — so the
    "notify the admins" half of the requirement is served by a number on surfaces administrators
    already open. This is that number, and it is a COUNT rather than a list so that the polling it
    invites stays cheap: ``AccessRoster_status_idx`` answers it from the index alone.
    """
    return await db.accessroster.count(where={"status": "PENDING"})


# ---------------------------------------------------------------------------------------------
# Admission bookkeeping
# ---------------------------------------------------------------------------------------------


async def ensure_admitted(
    email: Any,
    *,
    granted_role: str | None = None,
    full_name: str | None = None,
    notes: str | None = None,
    decided_by_id: str | None = None,
    joined_at: datetime | None = None,
) -> Any:
    """Put an address on the roster as ACTIVE, creating the row or promoting an existing one.

    THE ONE PLACE ADMISSION IS WRITTEN, called from the roster routes, from ``POST /users`` and from
    the admin seed script. That matters more than it looks: an administrator who creates an account
    through the users screen and does NOT get a roster row has created an account that cannot sign
    in, and they would find out about it from the new user, by telephone, a day later. Every path
    that mints a ``User`` goes through here.

    ``joinedAt`` IS STAMPED ONCE AND NEVER MOVED. It is the date of joining the platform; a person
    suspended in June and restored in September joined in March, and re-stamping on restore would
    quietly rewrite that. ``or existing.joinedAt`` — not the other way round — is what enforces it.
    """
    address = normalise_email(email)
    if not address:
        raise ValueError("An access roster entry needs an email address")

    now = datetime.now(UTC)
    role = role_value(granted_role) or None
    existing = await db.accessroster.find_unique(where={"email": address})
    if existing is not None:
        data: dict[str, Any] = {
            "status": ADMITTING_STATUS,
            "decidedAt": now,
            "decidedById": decided_by_id,
            "joinedAt": existing.joinedAt or joined_at or now,
        }
        if role:
            data["grantedRole"] = role
        if full_name is not None:
            data["fullName"] = clean_text(full_name, MAX_FULL_NAME_LENGTH)
        if notes is not None:
            data["notes"] = clean_text(notes, MAX_NOTES_LENGTH)
        return await db.accessroster.update(where={"email": address}, data=data)

    return await db.accessroster.create(
        data={
            "email": address,
            "status": ADMITTING_STATUS,
            "grantedRole": role or get_settings().default_signup_role,
            "fullName": clean_text(full_name, MAX_FULL_NAME_LENGTH),
            "notes": clean_text(notes, MAX_NOTES_LENGTH),
            "joinedAt": joined_at or now,
            "decidedAt": now,
            "decidedById": decided_by_id,
            # Zero, because nobody asked: this address was admitted by an administrator's decision
            # rather than by being turned away first. The distinction is visible on the roster.
            "requestCount": 0,
            "firstRequestedAt": now,
            "lastRequestedAt": now,
        }
    )


def clean_text(value: Any, limit: int) -> str | None:
    """Trim, cap, and turn an all-whitespace value into NULL rather than into ``" "``.

    A roster row whose ``fullName`` is a single space is not blank to any ``if row.fullName`` test in
    this codebase, so it renders as an empty cell that an admin cannot tell from a missing one.
    """
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None
    return text[:limit]


# ---------------------------------------------------------------------------------------------
# Serialisation
# ---------------------------------------------------------------------------------------------


def roster_payload(row: Any, *, account: Any = None) -> dict[str, Any]:
    """One roster row as the two admin clients read it.

    ``account`` is the matching ``User`` when one exists, and it is passed in rather than looked up
    here because the list endpoint resolves them in ONE query for the whole page. The account's live
    ``role`` is surfaced alongside ``grantedRole`` on purpose: those two disagree the moment somebody
    is promoted through the users screen, and an admin reading only the roster would otherwise be
    looking at a stale tier and believe it current. See ``AccessRoster.grantedRole`` in the schema.
    """
    return {
        "id": row.id,
        "email": row.email,
        "status": status_value(row.status),
        "grantedRole": role_value(row.grantedRole),
        "fullName": row.fullName,
        "notes": row.notes,
        "joinedAt": _iso(row.joinedAt),
        "firstSeenAt": _iso(row.firstSeenAt),
        "requestCount": row.requestCount,
        "firstRequestedAt": _iso(row.firstRequestedAt),
        "lastRequestedAt": _iso(row.lastRequestedAt),
        "decidedAt": _iso(row.decidedAt),
        "decidedById": row.decidedById,
        "createdAt": _iso(row.createdAt),
        "updatedAt": _iso(row.updatedAt),
        # Null when the address has been admitted but nobody has ever signed in with it — which is
        # exactly the state an admin needs to be able to see.
        "userId": getattr(account, "id", None),
        "accountRole": role_value(getattr(account, "role", None)) or None,
        "accountName": getattr(account, "name", None),
    }


def _iso(value: Any) -> str | None:
    return value.isoformat() if isinstance(value, datetime) else None

File: frontend/lib/accessRoster.ts
/**
 * The access roster — the client half of the sign-in gate.
 *
 * WHAT THE ROSTER IS. Until the gate shipped, this application refused nobody: any verified Google
 * address on earth got a `User` row and a bearer token, automatically, at the lowest tier. The
 * roster is the institution's answer to "who may sign in at all", and it is deliberately NOT the
 * `User` table — a row usually exists BEFORE the account does (an admin admits an address; the
 * account provisions itself the first time that person signs in) and it outlives the access
 * (suspending keeps the row, so an audit two years from now can still answer "who was admitted, and
 * when"). That is why every endpoint here is keyed by EMAIL and not by user id.
 *
 * THE PENDING HALF IS ALSO THE NOTIFICATION. There is no email and no push anywhere in this
 * codebase, so "the admins and master admins should get a notification to approve or reject the
 * user" is served by a count on surfaces admins already open (`GET /access-roster/pending-count`)
 * plus the queue itself (`GET /access-roster?status=PENDING`). Everything in this module exists to
 * feed one of those two.
 *
 * EVERY FUNCTION HERE IS `require_admin` ON THE SERVER (backend/app/api/routes/access_roster.py).
 * Mirror that with `canManageAccessRoster` before rendering a control — never after.
 */

import { apiFetch, buildQuery } from "@/lib/api";
import type { PageResult, UserRole } from "@/lib/types";

/**
 * The four statuses, mirroring the `AccessStatus` enum in `backend/prisma/schema.prisma`.
 *
 * Only ACTIVE admits. That is the whole security model and it is worth stating on the client too:
 * a PENDING row is created BY THE REFUSED CALLER, so "a row exists" and "this person may sign in"
 * are one clause apart, and any UI that treats the presence of a row as admission is describing an
 * authentication bypass to the admin reading it.
 */
export type AccessStatus = "PENDING" | "ACTIVE" | "REJECTED" | "SUSPENDED";

/** The statuses an admin may SET, matching `access_roster.DECIDABLE_STATUSES`. PENDING is absent. */
export const DECIDABLE_STATUSES: AccessStatus[] = ["ACTIVE", "REJECTED", "SUSPENDED"];

/** One roster row, exactly as `access_roster.roster_payload` serialises it. */
export type AccessRosterEntry = {
  id: string;
  email: string;
  status: AccessStatus;
  /** The tier this address gets when it is admitted. Not the live account role — see `accountRole`. */
  grantedRole: UserRole;
  /** ADMIN-WRITTEN ONLY. The sign-in path never stores a name a stranger supplied. */
  fullName: string | null;
  notes: string | null;
  /** The date of joining the platform. Stamped the FIRST time an address is admitted, never moved. */
  joinedAt: string | null;
  /** First successful sign-in. Null = admitted but never taken up — the invitation nobody opened. */
  firstSeenAt: string | null;
  /** How many times a refused sign-in bumped this row. Zero = an admin added it; nobody asked. */
  requestCount: number;
  firstRequestedAt: string | null;
  lastRequestedAt: string | null;
  decidedAt: string | null;
  decidedById: string | null;
  createdAt: string | null;
  updatedAt: string | null;
  /**
   * The account behind the address, when one exists. Resolved server-side in ONE query for the
   * whole page, which is why it arrives on the row instead of being looked up here.
   */
  userId: string | null;
  /**
   * The account's LIVE role, which disagrees with `grantedRole` the moment somebody is promoted
   * through the users screen. Both are shown on the roster on purpose: an admin reading only
   * `grantedRole` would be looking at a stale tier and believing it current.
   */
  accountRole: UserRole | null;
  accountName: string | null;
};

export type AccessRosterQuery = {
  page?: number;
  pageSize?: number;
  search?: string;
  /** Absent means EVERY status, which is the roster screen's default — see the page for why. */
  status?: AccessStatus;
};

/**
 * A page of the roster.
 *
 * REFUSED ROWS COME BACK UNLESS FILTERED OUT, and that default is the point of the screen: an admin
 * looking for somebody who says they cannot sign in needs to SEE the row that is refusing them. A
 * list that hid it would leave them re-adding an address the unique index then rejects with a 409,
 * with nothing on screen ever explaining why.
 */
export async function listAccessRoster(query: AccessRosterQuery = {}) {
  return apiFetch<PageResult<AccessRosterEntry>>(
    `/access-roster${buildQuery({
      page: query.page,
      pageSize: query.pageSize,
      search: query.search,
      status: query.status
    })}`
  );
}

/**
 * How many people are waiting on an administrator. THE NOTIFICATION, in its entirety.
 *
 * A bare count and not a list, deliberately: this is meant to be read by every admin surface, and
 * the server answers it out of an index without reading a row. Do not "improve" it into a call that
 * also returns the queue — the badge would become the most expensive query on the dashboard.
 */
export async function accessRosterPendingCount(): Promise<number> {
  const result = await apiFetch<{ pending: number }>("/access-roster/pending-count");
  return result.pending;
}

export type AccessRosterCreateBody = {
  email: string;
  fullName?: string | null;
  notes?: string | null;
  /** Absent means the bottom of the ladder — "all the users by default join as the lowest rung". */
  grantedRole?: UserRole | null;
  /** Defaults true on the server: "add somebody to the allow list" is what the endpoint is for. */
  isActive?: boolean;
};

/**
 * Put an address on the allow list. No account is required and none is created.
 *
 * A duplicate answers 409 with a sentence that NAMES the existing row and says restoring it is an
 * update rather than a second add. Show that sentence verbatim; it is the only thing that explains
 * why the add "did nothing".
 */
export async function addToAccessRoster(body: AccessRosterCreateBody) {
  return apiFetch<AccessRosterEntry>("/access-roster", {
    method: "POST",
    body: JSON.stringify(body)
  });
}

export type AccessRosterUpdateBody = {
  email?: string;
  status?: AccessStatus;
  grantedRole?: UserRole;
  fullName?: string | null;
  notes?: string | null;
};

/**
 * Approve, reject, restore or correct one entry.
 *
 * ABSENCE MEANS "LEAVE IT ALONE" — the server reads `exclude_unset`, so only send what changed. A
 * body that carried every field would let an admin fixing a typo in a note silently re-approve
 * somebody another admin had rejected.
 */
export async function updateAccessRosterEntry(id: string, body: AccessRosterUpdateBody) {
  return apiFetch<AccessRosterEntry>(`/access-roster/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body)
  });
}

/**
 * SUSPEND. Never a delete, whatever the HTTP verb says.
 *
 * `DELETE /access-roster/{id}` sets `status = SUSPENDED` and answers 200 with the suspended row, so
 * the entry stays on screen — dated, and one click from being restored. A real delete would put the
 * person straight back into the pending queue at their next sign-in, which is precisely the loop
 * REJECTED and SUSPENDED exist to break.
 */
export async function suspendAccessRosterEntry(id: string) {
  return apiFetch<AccessRosterEntry>(`/access-roster/${id}`, { method: "DELETE" });
}

/* ────────────────────────────────────────────────────────────────────────────
 * Shared wording.
 *
 * These strings are read by the roster page, the admin hub tile and the nav badge, and the Android
 * client says the same things in `ui/AccessRosterScreen.kt`. A status that reads "Pending" on one
 * surface and "Awaiting approval" on another is two features to the admin using both.
 * ──────────────────────────────────────────────────────────────────────────── */

export const ACCESS_STATUS_LABELS: Record<AccessStatus, string> = {
  PENDING: "Awaiting approval",
  ACTIVE: "May sign in",
  REJECTED: "Not approved",
  SUSPENDED: "Suspended"
};

/**
 * What each status MEANS for the person behind the address, in the admin's terms.
 *
 * Worded rather than coloured, because colour never carries meaning alone in this app — the
 * judgement has to survive colour-blindness, a greyscale print-out and forced-colours mode.
 */
export const ACCESS_STATUS_BLURBS: Record<AccessStatus, string> = {
  PENDING: "They tried to sign in and were turned away. Nothing happens until an admin decides.",
  ACTIVE: "They can sign in. Their account is created on first sign-in if it does not exist yet.",
  REJECTED: "They are refused, and retrying does not put them back in the queue.",
  SUSPENDED: "They had access and it was ended. Their sessions stop at the next request."
};

export function accessStatusLabel(status: AccessStatus | string): string {
  return ACCESS_STATUS_LABELS[status as AccessStatus] ?? String(status);
}

/**
 * The chip classes for a status.
 *
 * PENDING is amber rather than red: it is not a refusal an admin made, it is work waiting for one.
 * `amber-100`/`amber-800` and not `amber-50`/`amber-200` — only 100/500/800 are brand rungs in this
 * config, the rest deep-merge with stock Tailwind amber and do not pair.
 */
export function accessStatusChip(status: AccessStatus): string {
  switch (status) {
    case "ACTIVE":
      return "border-success-600/25 bg-success-100 text-success-600";
    case "PENDING":
      return "border-amber-500/30 bg-amber-100 text-amber-800";
    case "SUSPENDED":
    case "REJECTED":
    default:
      return "border-error-600/25 bg-error-100 text-error-600";
  }
}

/* ────────────────────────────────────────────────────────────────────────────
 * The refusal, as the sign-in page reads it.
 * ──────────────────────────────────────────────────────────────────────────── */

/**
 * The machine-readable half of a gated refusal: `ACCESS_PENDING`, `ACCESS_REJECTED` or
 * `ACCESS_SUSPENDED`, or null when the failure was anything else (a wrong password, a 500, no
 * network).
 *
 * WHY A CODE AND NOT A STRING MATCH. The server sends
 * `{"detail": {"code": "ACCESS_PENDING", "message": "…"}}` precisely so a client can branch on
 * something stable. Matching the English prose instead would break the first time somebody improves
 * a sentence, and the thing that would break is the sign-in page's ability to tell a person waiting
 * for an administrator apart from a person who mistyped their password — which is the entire ruling
 * this feature turns on.
 *
 * The MESSAGE is never composed here. `apiFetch` has already pulled `detail.message` out and put it
 * on `error.message`, and that sentence — written once, on the server, so that both clients and
 * both applications say the same thing — is the one shown to the person.
 */
export function accessRefusalCode(error: unknown): string | null {
  if (!error || typeof error !== "object") return null;
  const payload = (error as { payload?: unknown }).payload;
  if (!payload || typeof payload !== "object") return null;
  const detail = (payload as { detail?: unknown }).detail;
  if (!detail || typeof detail !== "object") return null;
  const code = (detail as { code?: unknown }).code;
  return typeof code === "string" && code.startsWith("ACCESS_") ? code : null;
}

/**
 * Is this refusal the "your request is with an administrator" one?
 *
 * Kept apart from the other two because it is the only one where the right thing for the person to
 * do is NOTHING — no retry, no password reset, no second attempt with another address. The sign-in
 * page renders it as a waiting state rather than as an error for exactly that reason.
 */
export function isAccessPending(error: unknown): boolean {
  return accessRefusalCode(error) === "ACCESS_PENDING";
}

/**
 * "Has this admitted address ever actually been used?" — the one question the roster exists to
 * answer that no other screen can.
 *
 * Never the word "never": a null `firstSeenAt` is a statement about an INVITATION, not about a
 * person, and an admin chasing five addresses added in March needs to be able to tell "they have
 * not opened it yet" from "we never let them in". An entry that is not admitted has no invitation
 * outstanding at all, so it says so instead.
 */
export function invitationLabel(entry: AccessRosterEntry): string {
  if (entry.firstSeenAt) return "Signed in";
  if (entry.status !== "ACTIVE") return "No access to take up";
  return "Not signed in yet";
}

/**
 * How the person behind a row got here — the difference between somebody an admin added and
 * somebody who asked.
 *
 * `requestCount` is 0 for a row an administrator created (nobody asked to be there) and rises by
 * one on every refused attempt. A rising count on a REJECTED row is how an admin notices somebody
 * who keeps trying and goes to talk to them, which is the whole reason a rejection stays a
 * rejection instead of re-opening.
 */
export function requestLabel(entry: AccessRosterEntry): string {
  if (!entry.requestCount) return "Added by an admin";
  if (entry.requestCount === 1) return "Asked once";
  return `Asked ${entry.requestCount} times`;
}
